### PR TITLE
SCP-4154 Pre-audit test report

### PR DESCRIPTION
**Pull request by bwbush**
[This document](https://github.com/input-output-hk/marlowe-cardano/blob/SCP-4154/marlowe/test/test-report.md) summarizes the state of Marlowe validator and on-chain testing in preparation for the Marlowe audit.

At this point, the following reviews are most valuable, but all critical comments are welcome!
- Document
    - Organization and coherence.
    - Level of detail.
 - Testing
    - Are there missing categories of tests that should be pursued?
    - Is more extensive testing required in particular areas?
    - Are there insufficiently mitigated issues that need attention (i.e., changes to Marlowe's Plutus validators)?

Proofreading for grammar, spelling, and broken links will take place when the document is finalized in a couple of weeks.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by bjornkihlberg** [comment]
*HINT*: One can read the document rendered [here](https://github.com/input-output-hk/marlowe-cardano/blob/SCP-4154/marlowe/test/test-report.md)

**Comment by bwbush** [comment]
I'd like someone to carefully check the reasoning in [the section on Marlowe failures caused by protocol limits](https://github.com/input-output-hk/marlowe-cardano/blob/SCP-4154/marlowe/test/test-report.md#funds-locked-or-input-blocked-due-to-exceeding-protocol-parameters).

**Review by hrajchert — APPROVED** [review]
Great work 🙌

I would maybe modify `.github/PULL_REQUEST_TEMPLATE.md` to include a warning that if we are modifying a test, this document is updated.

Also, it might be cool if we could distinguish the tests that are automatically tested in the CI for every PR versus the ones that needs to be run manually. Probably adding a section discussing the CI checks, assuming all test are automatic and adding a "manual" tag to the others.

**Review by palas — APPROVED** [review]
Very nice summary, thank you. I just wrote down a few small potential typos and so

File: marlowe/test/test-report.md [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+- The *Semantics Validator* is Plutus code that verifies that a transaction obeys the Marlowe DSL semantics along with the Cardano ledger rules. This validator resides in the `smallUntypedValidator` function that marshals Plutus types to call `computeTransaction` and checks constraints on input and output.
+- The *Payout Validator* is Plutus code that verifies that a transaction is properly authorized by a Marlowe role to withdraw funds paid by the Semantics Validator. This validator resides in the  `rolePayoutValidator`  function that simply verifies authorization.
+
+Marlowe's property-based and golden tests reside in the `marlowe-test` executable of the [marlowe](../) package of the [marlowe-cardano](https://github.com/input-output-hk/marlowe-cardano/) repository, while Marlowe's on-chain tests reside in the [`run-nonpab-test.sh`](../../marlowe-cli/run-nonpab-tests.sh) script of the [marlowe-cli](../../marlowe-cli/) package of the repository. The test suite `marlowe-actus-test` in the [marlowe-actus](../../marlowe-actus/) package heavily stresses the Haskell implementation of Marlowe semantics, too. Informal tests are scattered throughout the Marlowe repositories.
```
Comment (yveshauser) [inline marlowe/test/test-report.md:49]: Suggesting to remove the comment about ACTUS tests here. When separating the ACTUS code base into ACTUS core (no dependencies to Marlowe, resp. Cardano) and the Marlowe specific part, I put the ACTUS test-framework into ACTUS core only.
```suggestion
Marlowe's property-based and golden tests reside in the `marlowe-test` executable of the [marlowe](../) package of the [marlowe-cardano](https://github.com/input-output-hk/marlowe-cardano/) repository, while Marlowe's on-chain tests reside in the [`run-nonpab-test.sh`](../../marlowe-cli/run-nonpab-tests.sh) script of the [marlowe-cli](../../marlowe-cli/) package of the repository. Informal tests are scattered throughout the Marlowe repositories.
```

**Review by yveshauser — APPROVED** [review]
Thanks, really helpful!

File: marlowe/test/test-report.md [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+
+
+### Tests of Marlowe Semantics Implementation
+
+The property-based tests for the Haskell implementation of Marlowe semantics run are implemented in the [`Spec.Marlowe.Semantics`](Spec/Marlowe/Semantics/) modules. The table below lists categories of these tests and the name of each test.
+- The *golden tests* compare the output of the Haskell `computeTransaction` to a constraint-solver analysis of all possible valid paths through a reference set of Marlowe contracts; that analysis is independent of the Haskell implementation. Manually created sets of invalid inputs to the contract are also tested. Additional contracts could easily be added to this category of tests, but it is labor intensive to manually develop the invalid inputs.
+- The *holistic tests* verify the response of `computeTransaction` to specific types of invalid inputs and to specific side effects of applying inputs. The testing framework implements a monad that checks preconditions, invariants, and postconditions in the transaction computation that is being tested. This makes the test cases succinct and readable: only minimal effort would be required to include additional tests.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:76]: The tests are embedded in the `.hs` files for the `Spec.Marlowe.Semantics.Golden.*` modules. Generating the list of all paths through a contract takes too much time for the pangram contract for that to be done during the CI.

File: marlowe/test/test-report.md [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+    - `marlowe-actus-test` runs the ACTUS tests.
+- Scripted on-chain tests use `marlowe-cli` commands in `bash` scripts or its embedded testing DSL accessible via `marlowe-cli test script`.
+- Informal on- and off-chain tests use Jupyter notebooks.
+- Code coverage was checked using the GHC [HPC](https://wiki.haskell.org/Haskell_program_coverage) capability.
+
+
+## Summary of Testing History and Results
+
+The Marlowe validator implementation currently passes all property-based and on-chain tests. The testing code and process outlined above revealed several issues that were further scrutinized:
+1. The test for ["Constraint 4: No output to script on close"](../specification/ReadMe.md#constraint-4-no-output-to-script-on-close) of the Marlowe-Cardano specification originally failed. This failure was eliminated with a minor modification to the semantics validator.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:464]: See [slack discussion](https://input-output-rnd.slack.com/archives/GN03ZG44V/p1662745324280459) and the PR #257, which made the change.

File: marlowe/test/test-report.md [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+2. Testing revealed an ambiguity regarding how a payment of no value should be warned as `ReduceNonPositivePay` versus `ReducePartialPay`. In documentation we clarify that an intentional payment of zero results in the warning `ReduceNonPositivePay`, which means "the contract was meant to pay zero or less but that is impossible so it pays nothing", whereas a payment that happens to be zero because of insufficient funds results in the warning `ReducePartialPay`, which means "the contract is meant to pay more than there is, so it pays what there is".
+3. The `computeTransaction` function does not guard for the preconditions (1) that internal account balances are positive and (2) that duplicate `(Account, Token)` pairs are not present in the association map `accounts :: PlutusTx.Map (Account, Token) Integer`.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:466]: In some sense this is a general Plutus and EUTxO-model problem: there is never an on-chain check that sending funds to a script address includes datum that can be redeemed by the script. Thus, we will always be faced with the situation that the datum in a creation transaction might fail the Marlowe semantics preconditions. The _validator_ checks preconditions before sending to `computeTransaction`, which guards that `computeTransaction` will never receive a semantically invalid state.

[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:466]: I'll be interested to hear what the auditors' perspective is on this general issue.

File: marlowe/test/test-report.md [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+    1. Marlowe's semantics validator *guards* against the first case by failing to execute any transaction that begins with a non-positive balance in an internal account. *This causes permanent locking of funds in the contract.*
+    2. The semantics validator *does not guard* against duplicate accounts, in which circumstance an invalid initial state would be provided to `computeTransaction`, resulting in behavior that violates the semantics specification. *This potentially causes semantically invalid behavior of the contract.*
+4. A role-based payments to a party with a blank token name in a contract where no roles currency was defined (effectively making Ada the roles currency) can be withdrawn from the payout validator by anyone at all.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: > Should we disable this behaviour from both validators?

I think not. There will always be a possibility of off-chain code creating an invalid initial state, a nonsensical contract, etc., but we should allow that so long as Marlowe semantics behave as guaranteed. Adding on-chain protection against buggy off-chain code just increases the validator size.

File: marlowe/test/test-report.md [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+5. Slot number to POSIX millisecond time conversion from the node to the Plutus script context necessitates meticulous attention to time values in Marlowe contracts. In some cases, sub-second mismatches will cause a validation failure, but that can be mitigated by resubmitting the transaction with the correct slot-number validity interval.
+6. The POSIX-millisecond validity intervals in the Plutus script context may be open, closed, or half-open/half-closed, but the slot-number validity intervals at the node are always closed on the left and open on the right. This situation has a potential (similar to Issue 5 above) to make the temporal aspect of validation finicky.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:471]: > The semantics validator *does not guard* against duplicate accounts, in which circumstance an invalid initial state would be provided to `computeTransaction`, resulting in behavior that violates the semantics specification. *This potentially causes semantically invalid behavior of the contract.*

@palas, this has bothered me because we do guard against non-positive account balances as a precondition. We could add a `any . foldMap ... accounts` guard against this, but `foldMap` on an association list is somewhat expensive, plus we'd should probably do this for `choices` and `boundValues`, too.

File: marlowe/test/test-report.md [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+    4. A `Let` or `Choice` might increase the `Datum` size.
+2. Execution steps (`maxTxExecutionUnits.steps`)
+    1. A `Pay`, `If`, `Let`, or `Assert` might require an extensive evaluation of `Value` s and `Observations`, thus exceeding execution-step limits.
+    2. A `When` might include many `Case`s, so that evaluating the `Action`s within them (any of which might require extensive evaluation) exhausts the execution-step budget.
+    3. A `Close` is unlikely to exceed execution steps unless it triggers too many payments.
+    4. Because the execution of `Contract` terms is continued until a quiescent state is reached, the sequence of continuations might exceed step limits even though individual terms do not.
+3. Execution memory (`maxTxExecutionUnits.memory`)
+    1. The same considerations apply to execution memory as to execution steps.
+    2. This includes the sequence of `Contract` continuations mentioned above because execution memory is the measured as a *sum* not a *maximum* over the execution's steps.
+4. Minimum Ada in a UTxO (`utxoCostPerWord`)
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:495]: Yes, the validator allows extra Ada to satisfy min-Ada and we have extensively tested this on-chain.

File: marlowe/test/test-report.md [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+Our experience testing Marlowe's on-chain code highlights several areas of concern:
+1. *Gap between Isabelle proofs and Haskell implementation:*
+    1. The Haskell implementation was not created with a verified code generator.
+    2. Merkleization is absent from the Isabelle semantics.
+2. *Limitations of property-based testing:*
+    1. We use a limited set of golden test cases.
+    2. We currently do not have a technique for generating large numbers of nearly-valid transactions.
+    3. Correlations between `State`, `Contract`, `Input`, and `Environment` are extremely important for arbitrary-value generation that adequately explores realistic cases and edge cases.
+    4. Marlowe's test oracle is based on its PureScript implementation, not its Isabelle specification.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:524]: Thanks, I hadn't understood the relationship between the PureScript and Isabelle when I wrote this draft. I've corrected it.

File: marlowe/test/test-report.md [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+| [marlowe-cli/test/test-contract-for-differences.yaml](../../marlowe-cli/test/test-contract-for-differences.yaml) | A contract for differences, without an oracle. |
+| [marlowe-cli/test/test-coupon-bond-guaranteed.yaml](../../marlowe-cli/test/test-coupon-bond-guaranteed.yaml) | A guaranteed bond. |
+| [marlowe-cli/test/test-escrow-with-collateral.yaml](../../marlowe-cli/test/test-escrow-with-collateral.yaml) | An escrow contract that includes collateral. |
+| [marlowe-cli/test/test-escrow.yaml](../../marlowe-cli/test/test-escrow.yaml) | An escrow contract that does not include collateral. |
+| [marlowe-cli/test/test-simple.yaml](../../marlowe-cli/test/test-simple.yaml) | A simple notify-then-close contract. |
+| [marlowe-cli/test/test-swap-of-ada-and-dollar-tokens.yaml](../../marlowe-cli/test/test-swap-of-ada-and-dollar-tokens.yaml) | A swap of Ada tokens for USD tokens. |
+| [marlowe-cli/test/test-swap-of-ada-for-ada.yaml](../../marlowe-cli/test/test-swap-of-ada-for-ada.yaml) | A swap of Ada tokens for Ada tokens. |
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:422]: They are just fake `USD` tokens minted for the test and not intended to correspond to any "real life" tokens.

File: marlowe/test/test-report.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:80]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+- The *golden tests* compare the output of the Haskell `computeTransaction` to a constraint-solver analysis of all possible valid paths through a reference set of Marlowe contracts; that analysis is independent of the Haskell implementation. Manually created sets of invalid inputs to the contract are also tested. Additional contracts could easily be added to this category of tests, but it is labor intensive to manually develop the invalid inputs.
+- The *holistic tests* verify the response of `computeTransaction` to specific types of invalid inputs and to specific side effects of applying inputs. The testing framework implements a monad that checks preconditions, invariants, and postconditions in the transaction computation that is being tested. This makes the test cases succinct and readable: only minimal effort would be required to include additional tests.
+- The tests of *support functions* used by `computeTransaction` exhaustively cover each behavior of each function.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+- Finally, tests of the *suitability of data generation* (via [`Arbitrary`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck-Arbitrary.html) and [`Gen`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck-Gen.html#t:Gen)) ensure that the test inputs are neither too repetitious nor too unrealistic. We compute the entropy of the diversity of values generated by `arbitrary` to ensure that central, realistic, edge, simple, and complex cases are generated. Semantically interesting test cases should have meaningful and stressful correlation between `Environment`, `State`, `Contract`, and `Input`. To achieve this, we have implemented a `SemiArbitrary` typeclass that uses a carefully generated arbitrary  `Context` from which state, contract, input, and other values are drawn, but with some probability those values might be perturbed or completely random. Further work in generating correlated test values is warranted, since this would yield more demonstrably stressful test values.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+
+We have used Haskell code-coverage profiling ([HPC](https://wiki.haskell.org/Haskell_program_coverage)) to confirm that 100% of the logic paths in the [`Language.Marlowe.Core.V1.Semantics`](../src/Language/Marlowe/Core/V1/Semantics/) module are explored by the tests. There are, however, a couple of minor code fragments, such as the text of particular warning messages, that aren't evaluated.
+
+| Category | Subcategory | Test |
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md:88]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+
+
+### Tests of Plutus Functions Used by the Marlowe Implementation
+
+Marlowe's Haskell implementation relies on the Plutus Haskell functions behaving as advertised, but some Plutus documentation is terse. (There is at least one documented historical case where a misinterpretation of the behavior of a Plutus function was not detected during a Dapp audit, with the result that a Dapp containing a critical vulnerability was deployed on `mainnet` .) To supplement the preexisting tests of Plutus, we have written independent tests for the Plutus functions that Marlowe's validators use: see the table below. Particular emphasis was placed on `AssocMap` which is an association list that is used in Marlowe's Haskell as a map with unique keys. (Notably, `AssocMap` arguably has a misleading `Eq` instances and some other odd behavior.) Furthermore, the multiasset `Value`-comparison function, `(==)`, `leq` and `geq` are critical to Marlowe. Finally, inadequate documentation of `findDatum` and `findDatumHash` in Plutus V1 had formerly caused validation failures in Marlowe.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 14]
```diff
@@ -0,0 +1,543 @@
...
+                           -> Gen (PlutusTransaction PayoutTransaction)  -- ^ The generator.
+```
+Each test case modifies the arbitrary valid transaction using actions within `ArbitraryTransaction PayoutTransaction ()` that tailor the test case to the property being tested. Lenses are provided for simplifying the process of succinctly modifying the prospective transaction within its state monad. The "noising" switch provides the option to generate a transaction that just includes the bare-bones content for Marlowe versus a transaction that contains additional inputs, outputs, datum, and signatures that are not strictly necessary for the Marlowe transaction to be valid.
+
+
+#### Semantics Validator
+
+We use the aforementioned simulation technique to test properties constraining the semantics validator. The first tests run arbitrary valid transactions; the later tests run transactions that must fail for particular reasons. In three cases manual code inspection was used instead of a property-based tests: implementing those tests would have involved instrumenting the Plutus validator itself.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 15]
```diff
@@ -0,0 +1,543 @@
...
+| Constraint 11. Output contract | Invalid mismatch between contract and script output |
+| Constraint 12. Merkleized continuations | Mismatch between continuation and its merkleization |
+| Constraint 13. Positive balances | Invalid non-positive balance |
+| Constraint 14. Inputs authorized | Invalid missing authorization |
+| Constraint 15. Sufficient payment | Invalid insufficient payment |
+
+
+#### Payout Validator
+
+We use the aforementioned simulation technique to test properties constraining the payout validator. The first tests run arbitrary valid transactions; the later tests run transactions that must fail for particular reasons.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 16]
```diff
@@ -0,0 +1,543 @@
...
+
+Some of these limitations might be remedied by modifying the Cardano-Marlowe specification to allow more complex transactions (e.g., splitting payments so as to avoid the `maxValueSize` limit), but that would be at the expense of making the semantics validator quite a bit larger and more costly. Many of the limitations cannot be remedied by without a complete redesign of both the Plutus code for Marlowe and the `computeTransaction` implementation: it is likely that protocol parameters may limit *any* Marlowe limitation on Cardano. This highlights the importance of performing static analysis of a Marlowe contract *before* creating the contract on the blockchain. To verify that a contract will not experience these failure modes, one may need to attempt to execute off-chain every possible path the contract could take during its complete execution or equivalently prove that execution would not violate protocol limits.
+
+
+## Areas of Concern
+
+Our experience testing Marlowe's on-chain code highlights several areas of concern:
+1. *Gap between Isabelle proofs and Haskell implementation:*
+    1. The Haskell implementation was not created with a verified code generator.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 17]
```diff
@@ -0,0 +1,543 @@
...
+    2. Merkleization is absent from the Isabelle semantics.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: I added a small amount of text in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4

File: marlowe/test/test-report.md [hunk 18]
```diff
@@ -0,0 +1,543 @@
...
+| Value size       | `maxValueSize`               | locks |       |               |        |         | blocks    |          |          |
+
+Some of these limitations might be remedied by modifying the Cardano-Marlowe specification to allow more complex transactions (e.g., splitting payments so as to avoid the `maxValueSize` limit), but that would be at the expense of making the semantics validator quite a bit larger and more costly. Many of the limitations cannot be remedied by without a complete redesign of both the Plutus code for Marlowe and the `computeTransaction` implementation: it is likely that protocol parameters may limit *any* Marlowe limitation on Cardano. This highlights the importance of performing static analysis of a Marlowe contract *before* creating the contract on the blockchain. To verify that a contract will not experience these failure modes, one may need to attempt to execute off-chain every possible path the contract could take during its complete execution or equivalently prove that execution would not violate protocol limits.
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe/test/test-report.md]: Fixed in a6277158b2d823ce29bb35843df497cf54556bc4
